network/crl_test.go

```diff
@@ -0,0 +1,350 @@
+	// x509.ParseCRL() that is called inside CRL verifier is very fragile
+	// and will return error because of \n on start, that's why this func
+	// looks a little bit different from getTest*Cert()
+
+	// tests/ssl/crl.pem
+	return []byte(`-----BEGIN X509 CRL-----
+MIIB4zCBzDANBgkqhkiG9w0BAQsFADB0MQswCQYDVQQGEwJHQjEPMA0GA1UECAwG
+TG9uZG9uMQ8wDQYDVQQHDAZMb25kb24xGDAWBgNVBAoMD0dsb2JhbCBTZWN1cml0
+eTELMAkGA1UECwwCSVQxHDAaBgNVBAMME1Rlc3QgQ0EgY2VydGlmaWNhdGUXDTIw
+MTAyNjIzMjcwNloXDTIwMTEyNTIzMjcwNlowJzAlAhQH6gdHXIXsVUmzREIYCF6n
+ajAdThcNMjAxMDI2MjMyNzA2WjANBgkqhkiG9w0BAQsFAAOCAQEAKKJ56Ba7bOqE
+R8LFhfSwqqzGxVdPfDOu3d14cpRC3PJSLcyfLh+YA8Rkveed7Ee1RGaGjXrXTiZm
+p44CQcslUFoFGdONVTzTAamCphOiLh/LZB65d2dWo1WwOfsLwwaE+KTk+99+Ug/j
+EgiyuOwoyFqONlaDj4Z2PUC/mKmPuI1QJ9WB43oL8cDxPuOtJp0z2NEz0ZDl1pu6
+ghaEHc7lPDBmCOOGYneqEtfAi/wzMzKEdeXxV9s3MBWsXKfdrwsQc2pjLuH+K/wv
+aY3BPTUA7sZLJGFvsE2fh9yi1TsyKs4hsboS6FldZdXBnD/0ZqUnFy/KQ4t3LwH5
+L6zfD+wegQ==
+-----END X509 CRL-----`)
+}
+
+// Convert PEM-encoded certificate into DER and parse it into x509.Certificate
+func pemToX509Cert(data []byte) (*x509.Certificate, error) {
+	certDERBlock, _ := pem.Decode(data)
+	if certDERBlock == nil {
+		return nil, errors.New("Failed to decode PEM")
+	}
+
+	if certDERBlock.Type != "CERTIFICATE" {
+		return nil, errors.New("Decoded DER block was not a certificate")
+	}
+
+	cert, err := x509.ParseCertificate(certDERBlock.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	return cert, nil
+}
+
+func TestDefaultCRLClientHTTP(t *testing.T) {
+	crl := getTestCRL()
+	addr := "127.0.0.1:45318"
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/test_crl.pem", func(res http.ResponseWriter, req *http.Request) {
+		res.WriteHeader(200)
+		res.Header().Add("Content-Type", "application/pem-certificate-chain")
+		res.Write(crl)
+	})
+
+	httpServer := &http.Server{
+		Addr:    addr,
+		Handler: mux,
+	}
+	go func() {
+		httpServer.ListenAndServe()
+	}()
+	defer httpServer.Close()
+
+	// Give the server time to start, otherwise this test will fail sometimes
+	time.Sleep(time.Millisecond * 50)
+
+	crlClient := DefaultCRLClient{}
+
+	//
+	// Test with valid URI
+	//
+	uri := fmt.Sprintf("http://%s/test_crl.pem", addr)
+
+	fetchedCRL, err := crlClient.Fetch(uri)
+	if err != nil {
+		t.Fatalf("Unexpected error during reading %s: %v", uri, err)
+	}
+
+	if !bytes.Equal(crl, fetchedCRL) {
+		t.Fatalf("CRLClient returned mismatched data")
+	}
+
+	//
+	// Test with invalid URI (i.e. when server returns 404 not found or any other error)
+	//
+	uri = fmt.Sprintf("http://%s/wrong_filename.pem", addr)
+
+	fetchedCRL, err = crlClient.Fetch(uri)
+	if err == nil {
+		t.Fatalf("Unexpected success during reading CRL from wrong URI")
+	} else {
+		t.Logf("(Expected) fetch error: %v", err)
+	}
+}
+
+func TestDefaultCRLClientFile(t *testing.T) {
+	crl := getTestCRL()
+
+	file, err := ioutil.TempFile("/tmp", "go_test_*.crl")
+	if err != nil {
+		t.Fatalf("Cannot create temporary file to store CRL: %v", err)
+	}
```

Comment:
I guess an empty string instead of `"/tmp"` will be slightly better here, letting Go choose the default temporary directory for the system.